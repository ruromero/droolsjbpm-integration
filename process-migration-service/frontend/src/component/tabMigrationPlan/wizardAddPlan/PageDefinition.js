import React from "react";
import PropTypes from "prop-types";

import { ALERT_TYPE_ERROR } from "patternfly-react/dist/js/components/Alert/AlertConstants";

import Notification from "../../Notification";
import KieServerClient from "../../../clients/kieServerClient";
import ProcessSelector from "./ProcessSelector";
import { Form } from "patternfly-react/dist/js/components/Form";

export default class PageDefinition extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      errorMsg: "",
      containers: {}
    };
    this.loadProcessIds();
    this.props.onIsValid(
      this.validate(this.props.plan.source, this.props.plan.target) === ""
    );
    if (this.validateItem(this.props.plan.source)) {
      this.getDefinition(
        this.props.plan.source,
        this.props.setSourceDefinition
      );
    }
    if (this.validateItem(this.props.plan.target)) {
      this.getDefinition(
        this.props.plan.target,
        this.props.setTargetDefinition
      );
    }
  }

  loadProcessIds = () => {
    KieServerClient.getDefinitions(this.props.kieServerId).then(containers => {
      this.setState({
        containers
      });
    });
  };

  getDefinition = (definition, callbackFn) => {
    KieServerClient.getDefinition(
      this.props.kieServerId,
      definition.containerId,
      definition.processId
    )
      .then(res => {
        callbackFn(res);
      })
      .catch(err => {
        const status = err.response.status;
        let errorMsg;
        if (status === 404) {
          if (definition.containerId !== "" || definition.processId !== "") {
            errorMsg = `Process not found: ${definition.containerId}/${definition.processId}`;
          } else {
            errorMsg = "Provide a valid containerId and processId";
          }
          this.setState({ errorMsg });
          callbackFn("");
        } else {
          err.json().then(json => {
            this.setState({
              errorMsg: `${definition.containerId}/${definition.processId}: ${json.message.string}`
            });
            callbackFn("");
          });
        }
      });
  };

  validateItem = item => {
    const isFieldValid = field => {
      return field !== undefined && field !== "";
    };
    return (
      item !== undefined &&
      isFieldValid(item.containerId) &&
      isFieldValid(item.processId)
    );
  };

  validate = (source, target) => {
    const validSrc = this.validateItem(source);
    const validTgt = this.validateItem(target);
    if (!validSrc) {
      return "Provide a valid source";
    }
    if (!validTgt) {
      return "Provide a valid target";
    }
    if (
      source.containerId === target.containerId &&
      source.processId === target.processId
    ) {
      return "Source and target must be different";
    }
    return "";
  };

  onChangeSource = item => {
    const validationError = this.validate(item, this.props.plan.target);
    const isValidUpdate = validationError === "";
    this.props.onIsValid(isValidUpdate);
    this.props.onChangeSource(item.containerId, item.processId);
    if (this.validateItem(item)) {
      this.getDefinition(item, this.props.setSourceDefinition);
    }
    this.setState({ validationError });
  };

  onChangeTarget = item => {
    const validationError = this.validate(this.props.plan.source, item);
    const isValidUpdate = validationError === "";
    this.props.onIsValid(isValidUpdate);
    this.props.onChangeTarget(item.containerId, item.processId);
    if (this.validateItem(item)) {
      this.getDefinition(item, this.props.setTargetDefinition);
    }
    this.setState({ validationError });
  };

  render() {
    const notification = (
      <Notification
        type={ALERT_TYPE_ERROR}
        message={this.state.errorMsg}
        onDismiss={() => this.setState({ errorMsg: "" })}
      />
    );
    return (
      <Form horizontal>
        {this.state.errorMsg && notification}
        <ProcessSelector
          type="Source"
          options={this.state.containers}
          containerId={this.props.plan.source.containerId}
          processId={this.props.plan.source.processId}
          onChange={this.onChangeSource}
        />
        <ProcessSelector
          type="Target"
          options={this.state.containers}
          containerId={this.props.plan.target.containerId}
          processId={this.props.plan.target.processId}
          onChange={this.onChangeTarget}
          validationError={this.state.validationError}
        />
      </Form>
    );
  }
}

PageDefinition.propTypes = {
  kieServerId: PropTypes.string.isRequired,
  plan: PropTypes.object.isRequired,
  onIsValid: PropTypes.func.isRequired,
  onChangeSource: PropTypes.func.isRequired,
  onChangeTarget: PropTypes.func.isRequired,
  setSourceDefinition: PropTypes.func.isRequired,
  setTargetDefinition: PropTypes.func.isRequired
};
