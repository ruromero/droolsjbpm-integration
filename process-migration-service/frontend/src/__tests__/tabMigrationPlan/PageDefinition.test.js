import renderer from "react-test-renderer";
import React from "react";
import PageDefinition from "../../component/tabMigrationPlan/wizardAddPlan/PageDefinition";

test("PageDefinition renders correctly using snapshot", () => {
  const myMock = jest.fn();
  const kieServerId = "kie server id";
  const plan = {
    source: {
      containerId: "sourceContainerId"
    },
    target: {
      containerId: "targetContainerId"
    }
  };
  const tree = renderer
    .create(
      <PageDefinition
        kieServerId={kieServerId}
        plan={plan}
        onIsValid={myMock}
        onChangeSource={myMock}
        onChangeTarget={myMock}
        setSourceDefinition={myMock}
        setTargetDefinition={myMock}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
