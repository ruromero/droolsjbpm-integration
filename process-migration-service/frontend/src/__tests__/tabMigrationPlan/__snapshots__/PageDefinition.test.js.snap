// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageDefinition renders correctly using snapshot 1`] = `
<form
  className="form-horizontal"
>
  
  <div
    className="col-sm-12"
  >
    <h3>
      Source
    </h3>
  </div>
  <div
    className="required form-group"
  >
    <label
      className="col-sm-3 control-label"
      htmlFor="Sourcedropdown_containerid"
    >
      Container:
    </label>
    <div
      className="col-sm-9"
    >
      <div
        className="dropdown btn-group"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown-toggle btn btn-default"
          disabled={false}
          id="Source-ContainerDropdown"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button"
        >
          sourceContainerId
           
          <span
            className="caret"
          />
        </button>
        <ul
          aria-labelledby="Source-ContainerDropdown"
          className="dropdown-menu"
          role="menu"
        />
      </div>
    </div>
  </div>
  <div
    className="required form-group"
  >
    <label
      className="col-sm-3 control-label"
      htmlFor="Sourcedropdown_processid"
    >
      Process:
    </label>
    <div
      className="col-sm-9"
    >
      <div
        className="dropdown btn-group"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown-toggle btn btn-default"
          disabled={false}
          id="Source-ProcessDropdown"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button"
        >
          Select a process
           
          <span
            className="caret"
          />
        </button>
        <ul
          aria-labelledby="Source-ProcessDropdown"
          className="dropdown-menu"
          role="menu"
        />
      </div>
      
    </div>
  </div>
  <div
    className="col-sm-12"
  >
    <h3>
      Target
    </h3>
  </div>
  <div
    className="required form-group"
  >
    <label
      className="col-sm-3 control-label"
      htmlFor="Targetdropdown_containerid"
    >
      Container:
    </label>
    <div
      className="col-sm-9"
    >
      <div
        className="dropdown btn-group"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown-toggle btn btn-default"
          disabled={false}
          id="Target-ContainerDropdown"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button"
        >
          targetContainerId
           
          <span
            className="caret"
          />
        </button>
        <ul
          aria-labelledby="Target-ContainerDropdown"
          className="dropdown-menu"
          role="menu"
        />
      </div>
    </div>
  </div>
  <div
    className="required form-group"
  >
    <label
      className="col-sm-3 control-label"
      htmlFor="Targetdropdown_processid"
    >
      Process:
    </label>
    <div
      className="col-sm-9"
    >
      <div
        className="dropdown btn-group"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown-toggle btn btn-default"
          disabled={false}
          id="Target-ProcessDropdown"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button"
        >
          Select a process
           
          <span
            className="caret"
          />
        </button>
        <ul
          aria-labelledby="Target-ProcessDropdown"
          className="dropdown-menu"
          role="menu"
        />
      </div>
      
    </div>
  </div>
</form>
`;
